homeassistant/components/recorder/migration.py

```diff
@@ -2449,6 +2561,25 @@ def migrate_data(self, instance: Recorder) -> bool:
         """Migrate some data, returns True if migration is completed."""
         return self._migrate_data(instance)
 
+    def _ensure_index_exists(self, instance: Recorder) -> None:
+        """Ensure the index needed by the migration exists."""
+        if not self.index_to_drop:
+            return
+        table_name, index_name, base = self.index_to_drop
+        with session_scope(session=instance.get_session()) as session:
+            if get_index_by_name(session, table_name, index_name) is not None:
+                return
+        _LOGGER.warning(
+            (
+                "Data migration step '%s' needs index `%s` on table `%s`, but "
+                "it does not exist and will be added now"
+            ),
+            self.migration_id,
+            index_name,
+            table_name,
+        )
+        _create_index(instance, instance.get_session, table_name, index_name, base=base)
```

Comment:
Probably for a future PR, but it would be nice to have an end to end test that has a test database with a missing index is migrated as expected and the index was created.

Reply:
The problem is that the index is removed when the migration finishes, so I think we'll need some kind of mocking to check the index was created during the migration. If that's OK, we can make versions of the existing data migrator tests where we remove the index before finishing migration?

Reply:
> If that's OK, we can make versions of the existing data migrator tests where we remove the index before finishing migration?

Sounds good